File: android/app/src/main/kotlin/io/mynote/app/data/Entities.kt
package io.mynote.app.data

import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey
import io.mynote.core.Hlc

/**
 * Room mirrors of the core models.
 *
 * Storage types, not domain types: `:core` owns the domain so the sync engine
 * stays testable on the JVM. Each row keeps the `hlc` it was last written with —
 * that is what makes local merge decisions possible — and the `authorNode` from
 * inside that clock, so "everything this device owns" is an indexed query rather
 * than a scan.
 */

private fun nodeOf(hlc: String) = Hlc.decode(hlc)?.node.orEmpty()

@Entity(tableName = "notes", indices = [Index("parentId", "orderKey"), Index("authorNode")])
data class NoteRow(
    @PrimaryKey val id: String,
    val title: String = "",
    val icon: String? = null,
    val parentId: String? = null,
    val orderKey: String,
    val hlc: String,
    /** Device that wrote the current version; decoded from [hlc] on write. */
    val authorNode: String = nodeOf(hlc),
    val deleted: Boolean = false,
    val updatedAt: Long = System.currentTimeMillis(),
)

@Entity(tableName = "blocks", indices = [Index("noteId", "orderKey"), Index("authorNode")])
data class BlockRow(
    @PrimaryKey val id: String,
    val noteId: String,
    val parentId: String? = null,
    val orderKey: String,
    val type: String,
    /** JSON, matching the `content` field on the wire. */
    val content: String,
    val hlc: String,
    val authorNode: String = nodeOf(hlc),
    val deleted: Boolean = false,
    val updatedAt: Long = System.currentTimeMillis(),
)

@Entity(tableName = "themes", indices = [Index("authorNode")])
data class ThemeRow(
    @PrimaryKey val id: String,
    val name: String,
    val spec: String,
    val hlc: String,
    val authorNode: String = nodeOf(hlc),
    val deleted: Boolean = false,
)

/**
 * Which version of each other device's file we have already merged.
 *
 * Lets a sync skip a file that has not changed, instead of re-downloading every
 * device's whole backup on every pass.
 */
@Entity(tableName = "remote_versions")
data class RemoteVersionRow(
    @PrimaryKey val fileName: String,
    val version: String,
)

/** Sync bookkeeping. One row. */
@Entity(tableName = "sync_meta")
data class SyncMetaRow(
    @PrimaryKey val id: Int = 1,
    /** Newest clock we had when our own file was last uploaded successfully. */
    val lastUploadedHlc: String? = null,
    val lastSyncedAt: Long? = null,
    /** Which storage provider the notes currently belong to. */
    val connectedProvider: String? = null,
)

File: ios/MyNote/Sources/Persistence/Entities.swift
import Foundation
import SwiftData
import MyNoteCore

/// SwiftData mirrors of the core models.
///
/// These are storage types, not domain types: `MyNoteCore` owns the domain so
/// the sync engine stays testable without SwiftData. Each entity keeps the `hlc`
/// it was last written with, which is what makes local merge decisions possible.

@Model
final class NoteEntity {
    @Attribute(.unique) var id: String
    var title: String
    var icon: String?
    var parentId: String?
    var orderKey: String
    var hlc: String
    var deleted: Bool
    var updatedAt: Date

    init(id: String, title: String, icon: String?, parentId: String?,
         orderKey: String, hlc: String, deleted: Bool) {
        self.id = id
        self.title = title
        self.icon = icon
        self.parentId = parentId
        self.orderKey = orderKey
        self.hlc = hlc
        self.deleted = deleted
        self.updatedAt = .now
    }
}

@Model
final class BlockEntity {
    @Attribute(.unique) var id: String
    var noteId: String
    var parentId: String?
    var orderKey: String
    var type: String
    /// JSON, matching `blocks.content` on the server.
    var content: String
    var hlc: String
    var deleted: Bool
    var updatedAt: Date

    init(id: String, noteId: String, parentId: String?, orderKey: String,
         type: String, content: String, hlc: String, deleted: Bool) {
        self.id = id
        self.noteId = noteId
        self.parentId = parentId
        self.orderKey = orderKey
        self.type = type
        self.content = content
        self.hlc = hlc
        self.deleted = deleted
        self.updatedAt = .now
    }
}

@Model
final class ThemeEntity {
    @Attribute(.unique) var id: String
    var name: String
    var spec: String
    var hlc: String
    var deleted: Bool

    init(id: String, name: String, spec: String, hlc: String, deleted: Bool) {
        self.id = id
        self.name = name
        self.spec = spec
        self.hlc = hlc
        self.deleted = deleted
    }
}

/// A local edit waiting to reach the server.
///
/// Persisting the outbox is what makes the app genuinely offline-first: edits
/// made in airplane mode survive a force-quit and a reboot.
@Model
final class OutboxEntry {
    @Attribute(.unique) var key: String      // "<entity>:<id>"
    var entity: String
    var recordId: String
    var hlc: String
    var deleted: Bool
    var fieldsJSON: String
    var queuedAt: Date

    init(entity: String, recordId: String, hlc: String, deleted: Bool, fieldsJSON: String) {
        self.key = "\(entity):\(recordId)"
        self.entity = entity
        self.recordId = recordId
        self.hlc = hlc
        self.deleted = deleted
        self.fieldsJSON = fieldsJSON
        self.queuedAt = .now
    }
}

@Model
final class SyncState {
    @Attribute(.unique) var singleton: String = "state"
    var cursor: Int
    var lastSyncedAt: Date?

    init(cursor: Int = 0) {
        self.cursor = cursor
    }
}
